mm2src/lp_swap/recreate_swap_data.rs

```diff
@@ -0,0 +1,540 @@
+        uuid: started_event.uuid,
+        started_at: started_event.started_at,
+        maker_coin_start_block: started_event.maker_coin_start_block,
+        taker_coin_start_block: started_event.taker_coin_start_block,
+        // Don't set the fee since the value is used when we calculate locked by other swaps amount only.
+        maker_payment_trade_fee: None,
+        // Don't set the fee since the value is used when we calculate locked by other swaps amount only.
+        taker_payment_spend_trade_fee: None,
+        maker_coin_swap_contract_address: negotiated_event.maker_coin_swap_contract_addr.clone(),
+        taker_coin_swap_contract_address: negotiated_event.taker_coin_swap_contract_addr.clone(),
+    });
+    maker_swap.events.push(MakerSavedEvent {
+        timestamp: started_event_timestamp,
+        event: maker_started_event,
+    });
+
+    // Generate `Negotiated` event
+
+    let maker_negotiated_event = MakerSwapEvent::Negotiated(TakerNegotiationData {
+        taker_payment_locktime: started_event.taker_payment_lock,
+        taker_pubkey: started_event.my_persistent_pub,
+        maker_coin_swap_contract_addr: negotiated_event.maker_coin_swap_contract_addr,
+        taker_coin_swap_contract_addr: negotiated_event.taker_coin_swap_contract_addr,
+    });
+    maker_swap.events.push(MakerSavedEvent {
+        timestamp: negotiated_event_timestamp,
+        event: maker_negotiated_event,
+    });
+
+    // Then we can continue to process success Taker events.
+    let wait_refund_until = negotiated_event.maker_payment_locktime + 3700;
+    maker_swap
+        .events
+        .extend(convert_taker_to_maker_events(event_it, wait_refund_until));
+
+    Ok(maker_swap)
+}
+
+/// Converts `TakerSwapEvent` to `MakerSwapEvent`.
+/// Please note that this method ignores the [`TakerSwapEvent::Started`] and [`TakerSwapEvent::Negotiated`] events
+/// because they are used outside of this function to generate `MakerSwap` and the initial [`MakerSwapEvent::Started`] and [`MakerSwapEvent::Negotiated`] events.
+fn convert_taker_to_maker_events(
+    event_it: impl Iterator<Item = TakerSavedEvent>,
+    wait_refund_until: u64,
+) -> Vec<MakerSavedEvent> {
+    let mut events = Vec::new();
+    for TakerSavedEvent { event, timestamp } in event_it {
+        macro_rules! push_event {
+            ($maker_event:expr) => {
+                events.push(MakerSavedEvent {
+                    timestamp,
+                    event: $maker_event,
+                })
+            };
+        }
+
+        // This is used only if an error occurs.
+        let swap_error = SwapError {
+            error: format!("Origin Taker error event: {:?}", event),
+        };
+        match event {
+            // Even if we considered Taker fee as invalid, then we shouldn't have sent Maker payment.
+            TakerSwapEvent::TakerFeeSent(tx_ident) => push_event!(MakerSwapEvent::TakerFeeValidated(tx_ident)),
```

Comment:
We can be sure that the taker's fee was validated successfully only if the maker payment was sent afterward (`MakerPaymentReceived`). It seems we can ignore the `TakerFeeSent` because there is no analog for it like `TakerPaymentReceived`.

Reply:
I agree with you) I followed the logic that the taker fee is just a transaction that is not a critical part of the swap.
On the other hand, this change can complicate Taker event handling, since when we get `TakerSwapEvent::MakerPaymentReceived`, there is no info about the `TakerFee` transaction.
I can store `TransactionIdentifier` of the `TakerFee` as a variable for example if it's needed.

Reply:
> I can store TransactionIdentifier of the TakerFee as a variable for example if it's needed.

Seems like a solution. Could you do it this way, please?

Reply:
Sure, done